### ISSUE TITLE
[question] how to navigate with pop until or push based on history stack?

### ISSUE THREAD

**Issue by TatsuUkraine** [lulupointu/vrouter#75]
Sorry if this question is stupid) I just started to research various packages for routing, so I'm new with vrouter.

Question is next. For instance I have routes and pages like /users, /users/:id, /dashboard

Scenario one:
- I'm on /users
- navigate to users/:id
- finish with saving, want navigate back to users (simple pop)

Scenario two:
- I'm on dashboard
- navigate to /users/:id (if I click back, should be navigated to dashboard)
- finish with saving, want to navigate to users (need to make replace entire history with /users route)

So I wander, is there any option to make a navigate action after saving that will behave according to this rules:
- if /users exists in a history, it will do pop untill
- if /users doesn't exist in a history, it will erase all history and pushes a new route into the history

Thanks!

**Comment by lulupointu** [comment]
Hi, 

Two important things to understand:
1. 1 url = 1 history stack
2. pop on `someUrl` = push `someUrl`

### 1. 1 url = 1 history stack
If you want to have two different history stack: 
1. `UserScreen` | `UserWithIdScreen`
2. `DashboardScreen` | `UserWithIdScreen`

Then you need a different url for each of them. This is why relative url are usually recommended when stacking routes.
In your case you can have the following setup:
```dart
VRouter(
  initialUrl: '/users',
  routes: [
    VWidget(
      path: '/users',
      widget: UserScreen(),
      stackedRoutes: [VWidget(path: ':id', widget: UserWithIdScreen())],
    ),
    VWidget(
      path: '/dashboard',
      widget: DashboardScreen(),
      stackedRoutes: [VWidget(path: ':id', widget: UserWithIdScreen())],
    ),
  ],
);
```

As you can see here:
1. `UserScreen` | `UserWithIdScreen` : path = `/users/:id`
2. `DashboardScreen` | `UserWithIdScreen` : path = `/dashboard/:id`

Of course your can use the path you want. Even use absolute paths, the important thing it to have different paths for different stacks.

 ### 2. pop on `someUrl` = push `someUrl`

Now lets explore this, which you might not need in the end but is interesting especially if you are thinking in terms of `popUntil` which does not exist in `VRouter`.

<details>
<summary>
Let's keep the `VRouter` settings of **1**, and define our widgets like so:
</summary>

```dart
class UserScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('UserScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          SizedBox(height: 20),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/dashboard'),
            child: Text('Go to DashboardScreen'),
          ),
          Expanded(child: Container()),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/users/1'),
            child: Text('Go to UserWithIdScreen'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}

class DashboardScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('DashboardScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          SizedBox(height: 20),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/users'),
            child: Text('Go to UserScreen'),
          ),
          Expanded(child: Container()),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/dashboard/1'),
            child: Text('Go to UserWithIdScreen'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}

class UserWithIdScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('UserWithIdScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          Expanded(
            child: Center(
              child: Text('Press the back button to pop'),
            ),
          ),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/users'),
            child: Text('Save'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}
```
</details>

What I want you to try here, is to save from `/dashboard/:id` and from `/users/:id`. 
If you save from `/dashboard/:id`: it will navigate to `/users` because we push. Nothing special here.
If you save from `/users/:id`: it will **POP** to `/users` **EVEN THOUGH we pushed**.

If you pop however, it pops on the right `Screen`, because of **1** (different url for different stacks).

I hope you can understand the second point.

## I have a `VNester` for my scaffold but I want to use the stack route of `VNester`

Let's say you have a scaffold with a bottom navigation bar (which you certainly do) to navigate between `UserScreen` and `DashboardScreen`.
In that case you certainly want `UserWithIdScreen` to appear **on top** of the bottom navigation bar.

To achieve this you know (or would have come to know) that you have to use `VNester` and put `UserWithIdScreen` in `VNester.stackedRoutes`.

The issue here is that again, `VRouter` won't really know how to pop. You could use tricks but [the right way to do it is described here](https://vrouter.dev/guide/Advanced/Custom%20VRouteElement%20And%20Scaling)

<details>
<summary>
Here are those concepts applied to you case
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return VRouter(
      initialUrl: UsersRoute.users,
      routes: [
        UsersRoute(),
        DashboardRoute(),
      ],
    );
  }
}

class ScaffoldRouteElement extends VRouteElementBuilder {
  static final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

  final String path;
  final VRouteElement nestedRoute;
  final VRouteElement stackedRoute;

  ScaffoldRouteElement({
    required this.path,
    required this.nestedRoute,
    required this.stackedRoute,
  });

  @override
  List<VRouteElement> buildRoutes() => [
        VNester(
          key: ValueKey('ScaffoldRouteElement'),
          navigatorKey: navigatorKey,
          path: path,
          widgetBuilder: (child) => ScaffoldScreen(child),
          nestedRoutes: [nestedRoute],
          stackedRoutes: [stackedRoute],
        ),
      ];
}

class UsersRoute extends VRouteElementBuilder {
  static final String users = '/users';
  static final String _userId = users + '/:id';

  static void toUserId(BuildContext context, int id) => context.vRouter.push(users + '/$id');

  @override
  List<VRouteElement> buildRoutes() => [
        ScaffoldRouteElement(
          path: users,
          nestedRoute: VWidget(
            key: ValueKey(users),
            path: null,
            aliases: [_userId],
            widget: UsersScreen(),
          ),
          stackedRoute: VWidget(path: _userId, widget: UserWithIdScreen()),
        )
      ];
}

class DashboardRoute extends VRouteElementBuilder {
  static final String dashboard = '/dashboard';
  static final String _dashboardId = dashboard + '/:id';

  static void toUserId(BuildContext context, int id) =>
      context.vRouter.push(dashboard + '/$id');

  @override
  List<VRouteElement> buildRoutes() => [
        ScaffoldRouteElement(
          path: dashboard,
          nestedRoute: VWidget(
            key: ValueKey(dashboard),
            path: null,
            aliases: [_dashboardId],
            widget: DashboardScreen(),
          ),
          stackedRoute: VWidget(path: _dashboardId, widget: UserWithIdScreen()),
        )
      ];
}

class ScaffoldScreen extends StatelessWidget {
  final Widget child;

  const ScaffoldScreen(this.child);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: child,
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: context.vRouter.url!.contains(UsersRoute.users) ? 0 : 1,
        onTap: (value) =>
            context.vRouter.push(value == 0 ? UsersRoute.users : DashboardRoute.dashboard),
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.person), label: 'users'),
          BottomNavigationBarItem(icon: Icon(Icons.developer_board), label: 'dashboard'),
        ],
      ),
    );
  }
}

class UsersScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('UserScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          Expanded(child: Container()),
          ElevatedButton(
            onPressed: () => UsersRoute.toUserId(context, 1),
            child: Text('Go to UserWithIdScreen'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}

class DashboardScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('DashboardScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          Expanded(child: Container()),
          ElevatedButton(
            onPressed: () => DashboardRoute.toUserId(context, 1),
            child: Text('Go to UserWithIdScreen'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}

class UserWithIdScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('UserWithIdScreen')),
      body: Column(
        mainAxisSize: MainAxisSize.max,
        children: [
          Expanded(
            child: Center(
              child: Text('Press the back button to pop'),
            ),
          ),
          ElevatedButton(
            onPressed: () => context.vRouter.push('/users'),
            child: Text('Save'),
          ),
          SizedBox(height: 20),
        ],
      ),
    );
  }
}
```
</details>

However I **strongly suggest** that you read https://vrouter.dev/guide/Advanced/Custom%20VRouteElement%20And%20Scaling since many little things have to be understand for this to work well. Everything is explained in the link so I would no go over it again here.

I hope this helps you !

**Comment by TatsuUkraine** [comment]
Thanks for the explanation)

**Comment by lulupointu** [comment]
Your welcome. Feel free to close the issue if this answered your question (take your time if you need of course)

**Comment by Techno-Disaster** [comment]
### TLDR on rule 1: 
Have the following routes setup:
```
path: '/requests',
stackedRoutes: [path: '/requests/:type' ],
```
Navigating to requests and then to requests/type1 works fine and when you click the back button you are back at requests, but  if you do something like
```
path: '/requests/:requestid/:id',
```
You will go to /requests/type1/id1 , but the clicking the back button sends you back to the inital route `'/'` (or whatever is there above this stack, **NOT** `/requests/type1` or `requests`

@lulupointu LMK is this is correct, faced this issue today looks like this was the explanation

**Comment by lulupointu** [comment]
@Techno-Disaster this is right. Popping is relative to the route stack, not the url.

You can use `VPop` to redirect a pop of needed but in most cases this does a good job of doing what you want (in my experience).

**Comment by Techno-Disaster** [comment]
@lulupointu Sometimes pop and systemPop are handled differently, I'm doing `VRouter.of(context).push(context.vRouter.previousUrl);` which works well on pop but not systemPop. Just to be clear, I'm *not* following the 1 url = 1 history rule right now. But still is there anyway for systemPop to go to previousUrl too? VPopHandler did not handle it well too

**Comment by lulupointu** [comment]
Mmmm, this seems weird but I've done some changes to `pop` to make it compatible with `Navigator 1.0` recently so maybe I made a mistake that the tests did not catch.

Could you provide a small reproducible example?

**Comment by Techno-Disaster** [comment]
> Mmmm, this seems weird but I've done some changes to `pop` to make it compatible with `Navigator 1.0` recently so maybe I made a mistake that the tests did not catch.
> 
> Could you provide a small reproducible example?

I think this is expected? because as per your last comment, popping is based on the stackedRoutes and not previousUrl and as I mentioned I'm not following the 1 url = 1 history thingy.

**Comment by lulupointu** [comment]
Oh I understand what you mean ! Yes in that case it makes sense.

However I would like to understand what you mean by "VPopHandler" did not work.

Anyway if you provide a small example I think I'll be able to understand if all of this is really expected or not

**Comment by lulupointu** [comment]
@Techno-Disaster I'll close this for now since you seem to have fix your issue.

Please open a new one if you have a new question, no not to spam @TatsuUkraine
